File: 40-combination-sum-ii.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        self.ret = []
        self.helper(0, [], candidates, target, -1)
        return self.ret
        
    def helper(self, idx, arr, candidates, remain, last_selected):
        if idx >= len(candidates) or candidates[idx] > remain:
            return
        self.helper(idx+1, arr, candidates, remain, last_selected)
        if idx == 0 or candidates[idx] != candidates[idx-1] or last_selected == idx - 1:
            arr.append(candidates[idx])
            if remain - candidates[idx] == 0:
                self.ret.append(list(arr))
            self.helper(idx+1, arr, candidates, remain - candidates[idx], idx)
            arr.pop()

File: captainhcg/js/40-combination-sum-ii.js
var combinationSum2 = function(candidates, target) {
    candidates.sort(function(a, b){return a - b;})
    var arr = [];
    var ret = [];
    var helper = function(idx, lastIdx, remaining){
        if(idx >= candidates.length){
            return
        }
        var v = candidates[idx];
        if(v > remaining){
            return;
        }
        if(idx == 0 || candidates[idx] != candidates[idx-1] || idx === lastIdx + 1){
            arr.push(v)
            if(remaining - v === 0){
                ret.push(arr.slice(0))
            } else {
                helper(idx + 1, idx, remaining - v);
            }
            arr.pop()
        }
        helper(idx + 1, lastIdx, remaining)
    }
    helper(0, -1, target)
    return ret;
};
